ui/src/pages/QuickSetup/AccountSelectionModal.tsx

```diff
@@ -0,0 +1,262 @@
+      signatories: makeDamlSet([service.payload.provider, service.payload.customer]),
+      label: makeAccountName(accountInfo),
+      version: '0',
+    };
+
+    if (accountInfo.accountType === AccountType.REGULAR) {
+      const accountRequest: RequestOpenAccount = {
+        accountId,
+        observers: [serviceProvider],
+        ctrls: [service.payload.provider, service.payload.customer],
+      };
+      await ledger.exercise(Service.RequestOpenAccount, service.contractId, accountRequest);
+    } else {
+      const request: RequestOpenAllocationAccount = {
+        accountId,
+        observers: makeDamlSet<string>([]),
+        nominee: serviceProvider,
+      };
+      await ledger.exercise(Service.RequestOpenAllocationAccount, service.contractId, request);
+    }
+  };
+
+  const allocationAccountNeeded =
+    hasAllocationAccount && !allocationAccounts.length && !openAllocationAccountRequests.length;
+  const accountNeeded = hasRegularAccount && !accountNames.length && !openAccountRequests.length;
+
+  const accountExists = (accountInfo: AccountInfo) => {
+    const accountName = makeAccountName(accountInfo);
+    if (accountInfo.accountType === AccountType.REGULAR) {
+      return !!accounts.find(a => a.id.label === accountName);
+    } else {
+      return allocationAccounts.find(a => a.id.label === accountName);
+    }
+  };
+
+  const accountRequestExists = (accountInfo: AccountInfo) => {
+    const accountName = makeAccountName(accountInfo);
+    if (accountInfo.accountType === AccountType.REGULAR) {
+      return !!openAccountRequests.find(rq => rq.payload.accountId.label === accountName);
+    } else {
+      return !!openAllocationAccountRequests.find(rq => rq.payload.accountId.label === accountName);
+    }
+  };
+
+  const uuid = uuidv4();
+  const accountRequestPrepend = `${uuid}REQUEST`;
+  return (
+    <Modal
+      as={Form}
+      open={open}
+      onSubmit={() => {
+        const accts = _.mapValues(accountInfos, (accountInfo, k) => {
+          const account =
+            accountInfo.accountType === AccountType.REGULAR
+              ? accounts.find(a => a.id.label === accountNamesState[k])
+              : allocationAccounts.find(a => a.id.label === accountNamesState[k]);
+          return account;
+        });
+        onFinish(accts);
+        setAccountNamesState(emptyNamesState);
+        setOpen(false);
+      }}
+    >
+      <Modal.Header as="h2">{`Select Accounts for ${getName(party)} requesting from ${getName(
+        serviceProvider || ''
+      )}`}</Modal.Header>
+      <Modal.Content>
+        {!custodyServices.length && (allocationAccountNeeded || accountNeeded) ? (
+          <>This party must have at least one Custody service</>
+        ) : (
+          !!serviceProvider && (
+            <div>
+              {_.toPairs(accountInfos).map(([k, accountInfo]) => {
+                const providerOptions =
+                  accountExists(accountInfo) || accountRequestExists(accountInfo)
```

Comment:
its would be more clear to make this its own component (`<ProviderOption accountInfo={accountInfo} etc../>`, and pass props accordingly so the variable logic is not in the render function

Reply:
definitely much cleaner like this and with the `onChange` as its own function!